# **Mectrics for Neural Network**

In [3]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
import tensorflow.keras as keras
from tensorflow.keras.datasets import mnist

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score

In [4]:
# Load the MNIST dataset
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Data preparation
x_train = x_train.reshape(-1, 28*28) / 255.0
x_test = x_test.reshape(-1, 28*28) / 255.0

11490434/11490434 ━━━━━━━━━━━━━━━━━━━━ 2s 0us/step


In [5]:
# Logistic Regression Model Training
logistic_model = LogisticRegression(max_iter=1000)
logistic_model.fit(x_train, y_train)

LogisticRegression(max_iter=1000)

In [6]:
# Model Evaluation
y_pred_logistic = logistic_model.predict(x_test)

# calculate matrix and accuracy
conf_matrix_logistic = confusion_matrix(y_test, y_pred_logistic)
accuracy_logistic = accuracy_score(y_test, y_pred_logistic)
precision_logistic = precision_score(y_test, y_pred_logistic, average='macro')
recall_logistic = recall_score(y_test, y_pred_logistic, average='macro')
f1_logistic = f1_score(y_test, y_pred_logistic, average='macro')

# Print the results
print("Logistic Regression Model")
print("Confusion Matrix:")
print(conf_matrix_logistic)
print("Accuracy:", accuracy_logistic)
print("Precision:", precision_logistic)
print("Recall:", recall_logistic)
print("F1 Score:", f1_logistic)

Logistic Regression Model
Confusion Matrix:
[[ 957    0    1    4    1    9    4    3    1    0]
 [   0 1110    5    2    0    2    3    2   11    0]
 [   6    9  930   16   10    3   12    9   33    4]
 [   4    1   18  920    1   24    2   10   22    8]
 [   1    3    7    3  922    0    6    4    6   30]
 [   9    2    3   34    8  779   15    7   31    4]
 [   8    3    8    2    7   17  910    2    1    0]
 [   1    7   23    7    6    1    0  949    2   32]
 [  10   11    6   21    8   28   14    9  855   12]
 [   9    7    1    9   21    7    0   21    8  926]]
Accuracy: 0.9258
Precision: 0.9248160206554281
Recall: 0.924738776208031
F1 Score: 0.9247146011942394


### การคำนวณ TP, TN, FP, FN จาก Confusion Matrix ช่วยบอกอะไรเกี่ยวกับความแม่นยำของโมเดลนี้?

Confusion Matrix เป็นตารางที่แสดงผลการพยากรณ์ของโมเดลเปรียบเทียบกับค่าจริง โดยประกอบด้วย:

- **True Positive (TP):** จำนวนกรณีที่โมเดลพยากรณ์ถูกต้องว่าเป็นคลาสนั้นๆ
- **True Negative (TN):** จำนวนกรณีที่โมเดลพยากรณ์ถูกต้องว่าไม่เป็นคลาสนั้นๆ
- **False Positive (FP):** จำนวนกรณีที่โมเดลพยากรณ์ว่าเป็นคลาสนั้นๆ แต่จริงๆ ไม่เป็น
- **False Negative (FN):** จำนวนกรณีที่โมเดลพยากรณ์ว่าไม่เป็นคลาสนั้นๆ แต่จริงๆ เป็น

การคำนวณค่าเหล่านี้จาก Confusion Matrix ช่วยให้เราเข้าใจรายละเอียดเกี่ยวกับความแม่นยำของโมเดลได้ลึกซึ้งขึ้น
ไม่เพียงแค่ดูจากค่า Accuracy เท่านั้น แต่ยังสามารถวิเคราะห์การทำงานของโมเดลในแต่ละคลาสได้ว่าโมเดลมีความสามารถในการจำแนกแต่ละคลาสได้ดีเพียงใด
มีการพยากรณ์ผิดพลาดในลักษณะใดบ้าง

### วิเคราะห์ผลลัพธ์ของเมตริกต่าง ๆ ที่ได้จากโมเดล

จากข้อมูลที่ให้มา โมเดล Logistic Regression มีเมตริกต่าง ๆ ดังนี้:

- **Accuracy (ความถูกต้อง):** 0.9258  
  แสดงว่าโมเดลสามารถพยากรณ์ได้ถูกต้องใน 92.58% ของกรณีทั้งหมด 
  เป็นตัวชี้วัดทั่วไปที่บอกถึงความสามารถโดยรวมของโมเดล

- **Precision (ความแม่นยำ):** 0.9248  
  แสดงว่าเมื่อโมเดลพยากรณ์ว่าเป็นคลาสใดคลาสหนึ่ง มักจะถูกต้องใน 92.48% ของกรณีนั้นๆ 
  หมายความว่าโมเดลมีอัตราการพยากรณ์ผิดประเภทเป็น False Positive ต่ำ

- **Recall (ความสามารถในการดึงข้อมูล):** 0.9247  
  แสดงว่าโมเดลสามารถจับคลาสที่แท้จริงได้ใน 92.47% ของกรณีทั้งหมด 
  หมายความว่าโมเดลมีอัตราการพลาดประเภทเป็น False Negative ต่ำ

- **F1 Score:** 0.9247  
  เป็นค่าเฉลี่ยฮาร์มอนิกของ Precision และ Recall 
  ซึ่งให้ภาพรวมที่สมดุลระหว่างความแม่นยำและความสามารถในการดึงข้อมูล 
  โมเดลมี F1 Score ที่สูง แสดงว่าทั้ง Precision และ Recall อยู่ในระดับดี

จากเมตริกเหล่านี้ สามารถสรุปได้ว่าโมเดล Logistic Regression นี้มีประสิทธิภาพสูงในการจำแนกภาพตัวเลข 
มีความสมดุลระหว่างการพยากรณ์ที่ถูกต้องและการจับคลาสที่แท้จริงได้ดี

### ระบุว่าเมตริกใดมีความสำคัญที่สุดในบริบทของการจำแนกภาพตัวเลข และให้เหตุผลว่าทำไมเมตริกนี้จึงสำคัญ

ในบริบทของการจำแนกภาพตัวเลข เมตริก **F1 Score** ถือว่าเป็นเมตริกที่สำคัญที่สุด เนื่องจาก:

1. **สมดุลระหว่าง Precision และ Recall:** ในการจำแนกภาพตัวเลข การมี Precision 
    สูงหมายถึงการพยากรณ์ที่ถูกต้องเมื่อโมเดลระบุเลขหนึ่งๆ และ Recall สูงหมายถึงการจับเลขนั้นๆ ได้ครบถ้วน 
    การใช้ F1 Score ซึ่งเป็นการรวมค่าทั้งสองเข้าด้วยกัน ช่วยให้เรามองเห็นประสิทธิภาพโดยรวมของโมเดลได้ดียิ่งขึ้น

2. **การจัดการกับข้อมูลไม่สมดุล:** หากข้อมูลมีการกระจายไม่สมดุล (เช่น บางเลขมีจำนวนมากกว่าเลขอื่น) 
    F1 Score จะช่วยให้การประเมินโมเดลไม่ถูกเบี่ยงเบนไปในทางใดทางหนึ่ง 
    ซึ่งเป็นสิ่งสำคัญในการจำแนกภาพตัวเลขที่อาจมีคลาสที่แตกต่างกันในจำนวนตัวอย่าง

3. **การประเมินแบบครอบคลุม:** เนื่องจาก F1 Score ผสมผสานทั้ง Precision และ Recall เข้าด้วยกัน 
จึงเป็นตัวชี้วัดที่ครอบคลุมและเหมาะสมในการประเมินโมเดลที่ต้องการความแม่นยำสูงและความครบถ้วนในการจับคลาสที่แท้จริง

ดังนั้น F1 Score จึงเป็นเมตริกที่สำคัญที่สุดในการจำแนกภาพตัวเลข 
เนื่องจากให้ภาพรวมที่สมดุลและครอบคลุมถึงความสามารถของโมเดลทั้งในด้านการพยากรณ์ที่ถูกต้องและการจับข้อมูลที่แท้จริง
